VM/rel2vm_params.py

```diff
@@ -0,0 +1,1105 @@
+    if math.floor(land1) == 0:
+        logger.info("Land coverage is less than 1%. Not creating VM")
+        return None
+
+    # zlen is independent from xlen and ylen
+    zlen = round(get_max_depth(faultprop.Mw, srf_meta["dbottom"]) / hh) * hh
+    logger.debug(f"zlen set to {zlen}")
+
+    if xlen1 == 0 or ylen1 == 0 or zlen == 0:
+        logger.debug(
+            f"All xlen={xlen1} ylen={ylen1} zlen={zlen} should be non-zero. Not creating VM"
+        )
+        return None
+
+    # modified sim time
+    vm_corners = np.asarray([c1, c2, c3, c4])
+    initial_time = get_sim_duration(
+        vm_corners,
+        np.concatenate(srf_meta["corners"], axis=0),
+        ds_multiplier,
+        fault_depth,
+        logger=logger,
+    )
+    sim_duration = (initial_time // dt) * dt
+    logger.debug(
+        f"Unrounded sim duration: {initial_time}. Rounded sim duration: {sim_duration}"
+    )
+
+    # optimisation results
+    return {
+        "name": srf_meta["name"],
+        "mag": faultprop.Mw,
+        "dbottom": srf_meta["dbottom"],
+        "zlen": zlen,
+        "sim_time": sim_duration,
+        "xlen": xlen0,
+        "ylen": ylen0,
+        "land": land0,
+        "zlen_mod": zlen,
+        "sim_time_mod": sim_duration,
+        "xlen_mod": xlen1,
+        "ylen_mod": ylen1,
+        "land_mod": land1,
+        "origin": origin,
+        "bearing": bearing,
+        "adjusted": optimise,
+        "plot_region": plot_region,
+        "path": "{:.6f}\t{:.6f}\n{:.6f}\t{:.6f}\n{:.6f}\t{:.6f}\n{:.6f}\t{:.6f}\n".format(
+            o1[0], o1[1], o2[0], o2[1], o3[0], o3[1], o4[0], o4[1]
+        ),
+        "path_mod": "{:.6f}\t{:.6f}\n{:.6f}\t{:.6f}\n{:.6f}\t{:.6f}\n{:.6f}\t{:.6f}\n".format(
+            c1[0], c1[1], c2[0], c2[1], c3[0], c3[1], c4[0], c4[1]
+        ),
+    }
+
+
+def main(
+    srf_meta: dict,
+    ds_multiplier: float,
+    dt: float,
+    hh: float,
+    min_vs: float,
+    outdir: Path,
+    pgv: float,
+    vm_topo: str,
+    vm_version: str,
+    deep_rupture: bool = False,
+    target_land_coverage: float = 99.0,
+    optimise: bool = True,
+    plot_enabled: bool = True,
+    logger: Logger = qclogging.get_basic_logger(),
+):
+    """
+    Orchestrates conversion from rel CSV file to vm_params.yaml
+
+    Parameters
+    ----------
+    srf_meta : Data extracted from REL csv file
+    ds_multiplier :
+    dt :
+    hh :
+    min_vs :
+    outdir : Directory where output files are eventually saved (eg. ..../Data/VMs/Hossack)
+    pgv :
+    vm_topo :
+    vm_version :
+    deep_rupture :  If true, continue even if the rupture is too deep.
+    target_land_coverage : Land coverage level (%) that triggers optimisation if not met.
+    optimise : Performs area optimisation if set true
+    plot_enabled : If true, plot the vm domain
+    logger:
+    """
+
+    # temp directory for current process
+    with TemporaryDirectory(prefix=f"_tmp_{srf_meta['name']}_", dir=outdir) as ptemp:
```

Comment:
Better variable name than ptemp? I'm not even sure what ptemp stands for?

Reply:
> Last thing I need would be a comparison between old code and this to make sure the VM parameters are similar / the same.

Good point. The new code works out corners, dbottom, and hdepth from the data it read from the rel csv, while old srfinfo2vm reads these data directly from HDF5 (which loses some precision of floating point data when saved), for example,
```
   ATTRIBUTE "corners" {
      DATATYPE  H5T_IEEE_F64LE
      DATASPACE  SIMPLE { ( 1, 4, 2 ) / ( 1, 4, 2 ) }
      DATA {
      (0,0,0): 176.25, -38.3302,
      (0,1,0): 176.22, -38.3498,
      (0,2,0): 176.181, -38.3222,
      (0,3,0): 176.211, -38.3025
      }
   }
```
So there is slight difference, especially to centroidDepth and MODEL_LAT and MODEL_LON 
(I can't figure out which one is old or new)
```
centroidDepth: 5.399587694214754			      |	centroidDepth: 5.36477
MODEL_LAT: -38.32617169848253				      |	MODEL_LAT: -38.32618573280984
MODEL_LON: 176.2154288572724				      |	MODEL_LON: 176.21539680790352
```

 I observed the difference and discussed with @jpa198  during the early stage of development and we decided this disparity is acceptable.

Reply:
and ptemp is renamed to temp_dir

Reply:
Sweet those small differences are acceptable. I'd like to see a direct comparison for the VM dimensions and sim_time for a couple of scenarios. (Hossack, Moonshine and Alpine are our staples if you want to use them)

Also, I hope that vm_gen isn't working out h_depth/dbottom those are fault parameters.

Reply:
It doesn't look like dbottom is saved for type 2s, we should probably change that

Reply:
Enlighten me - I couldn't find hdepth and dbottom from rel.csv Are they hiding here in disguise?
```
type,magnitude,moment,name,fault_type,tect_type,rake,dip,dtop,dbottom,length,plane_count,slip_rate,dip_dir,shypo,dhypo,clon_subfault_0,clat_subfault_0,type_subfault_0,magnitude_subfault_0,moment_subfault_0,name_subfault_0,strike_subfault_0,rake_subfault_0,dip_subfault_0,dtop_subfault_0,dbottom_subfault_0,length_subfault_0,width_subfault_0,dip_dir_subfault_0,dt,seed,genslip_version,sdrop,srfgen_seed
4,5.333953679644069,1.995671870606849e+24,Hossack_REL03,NORMAL_FAULTING,VOLCANIC,-101.51295394380425,60.0,0.0,8.0,,1,0.12,312.0,-0.14046251262771425,4.745585501477487,176.23500207053635,-38.340000955337224,0,5.500059423932996,1.995671870606849e+24,Hossack,229.62723233986827,-90.0,60.0,0.0,8.0,3.4,9.2,312.0,0.005,1878532741,5.4.2,41.52536628686057,939956369
```

Reply:
![image](https://user-images.githubusercontent.com/25143301/122862035-9be3d980-d374-11eb-85c0-b01ea4f50da6.png)
 Found it for you.
hdepth is only used if dtop is not available (aka a point source) In this case, dbottom should also be hdepth. Ideally, we wouldn't make these assumptions and the rel_csv is complete. @jpa198

Reply:
you are right - dbottom is in rel. 

I glanced at the code, and thought dbottom was computed (I originally said, "The new code works out corners, dbottom, and hdepth"),  but it in fact, it uses the value taken from rel. It's just that dbottom is retrieved as a by-product when calling get_corners_dbottom function.
```
    corners, dbottom = get_corners_dbottom(planes, dip_dir=rel_meta["dip_dir"].loc[0])
...
    dbottom = rel_meta["dbottom"].loc[0]

    hdepth = (
        np.round(rel_meta["dhypo"].loc[0], decimals=1)
        * np.sin(np.radians(rel_meta["dip"].loc[0]))
        + dtop
    )
```
get_corners_dbottom() is the same function that create_info_file() uses when creating srfinfo, but as "planes" have already lost its precision, dbottom returned by the function is slightly different from dbottom retrieved from rel_meta. (7.96 vs 8.0). I could have just called the function this way to ignore the returned dbottom.
```
corners, _ = get_corners_dbottom(planes, dip_dir=rel_meta["dip_dir"].loc[0])
```
rel_csv is unfortunately not complete, and regardless of srf type, both hdepth and dbottom are needed for the following steps (hdepth is written in vm_params.yaml and dbottom is used to compute zlen. I guess what you meant by 

> hdepth is only used if dtop is not available (aka a point source) In this case, dbottom should also be hdepth. Ideally, we wouldn't make these assumptions and the rel_csv is complete. @jpa198

is you are ok with what it is.

Reply:
I've had a cursory look. hdepth in vm_params doesn't seem to be used. So it should probably be removed as it isn't a VM parameter. 

I'm okay with the numbers being slightly different - what does matter though is making sure that small changes in inputs doesn't lead to large changes in the outputs hence why I want to see outputs.

After this conversation, I reckon it'd be best to see that this new code works for SRF types 1-4 and ad-hoc bugfixes to the rel.csv generation made as necessary (out of scope but if it doesn't work, it's good to know now).

Reply:
hdepth is not used in VM generation, but it is used in optimise_vm_params() to determine fault_depth. I'd like to be careful here before removing it.

I have done type1,2,4 tests (@jpa198  reckons type 3 is not necessary), and the latest code works for all cases.

Reply:
Can we please see numeric comparisons (for VM dimensions and simulation time) between old and new for type 1,2,4?

Reply:
Also, hdepth is used internally for rel2vm_params but shouldn't be used to actually generate the VM - and hence shouldn't be in the vm yaml IMO.

Reply:
vm_params.yaml is the only "readable" file that gives clue about the nature of the 3d vm files - even if it's not used for VM generation, I am reluctant to delete it. If it is a group consensus, we could delete other fields too.
I checked slurm_gm_workflow and it appears "output_directory" and "extracted_slice_parameters_directory" and code can go. In fact, code has a hard-coded "rt" which stands for "real-time" that somehow survived without a good reason.
@jpa198 What do you think?

Reply:
srfinfo2vm: for 2016p666385
```
mag: 5.3
centroidDepth: 7.0
MODEL_LAT: -37.91452293640712
MODEL_LON: 178.65619142526242
MODEL_ROT: 342.2236566127869
hh: 0.4
min_vs: 0.5
model_version: '1.65'
topo_type: BULLDOZED
output_directory: output
extracted_slice_parameters_directory: SliceParametersNZ/SliceParametersExtracted.txt
code: rt
extent_x: 128.4
extent_y: 138.8
extent_zmax: 32.800000000000004
extent_zmin: 0.0
sim_duration: 74.75
flo: 0.25
nx: 321
ny: 347
nz: 82
sufx: _rt01-h0.400

```
Type 1 2016p666385
```
mag: 5.3
centroidDepth: 7.0
MODEL_LAT: -37.90578440383951
MODEL_LON: 178.6533328489272
MODEL_ROT: 342.2253845879213
hh: 0.4
min_vs: 0.5
model_version: '2.06'
topo_type: BULLDOZED
output_directory: output
extracted_slice_parameters_directory: SliceParametersNZ/SliceParametersExtracted.txt
code: rt
extent_x: 128.0
extent_y: 136.40000000000003
extent_zmax: 32.0
extent_zmin: 0.0
sim_duration: 74.28
flo: 0.25
nx: 320
ny: 341
nz: 80
sufx: _rt01-h0.400

```
Type 2 2016p666385
```
mag: 5.3
centroidDepth: 6.966831131657108
MODEL_LAT: -37.91385094151386
MODEL_LON: 178.65324780949922
MODEL_ROT: 342.22365112347444
hh: 0.4
min_vs: 0.5
model_version: '2.06'
topo_type: BULLDOZED
output_directory: output
extracted_slice_parameters_directory: SliceParametersNZ/SliceParametersExtracted.txt
code: rt
extent_x: 128.4
extent_y: 138.8
extent_zmax: 32.800000000000004
extent_zmin: 0.0
sim_duration: 74.74
flo: 0.25
nx: 321
ny: 347
nz: 82
sufx: _rt01-h0.400

Reply:
Hossack (srfinfo2vm)
```
mag: 5.640075317543397
centroidDepth: 4.50641
MODEL_LAT: -38.32092325184884
MODEL_LON: 176.20795517540762
MODEL_ROT: 1.0
hh: 0.4
min_vs: 0.5
model_version: '1.65'
topo_type: BULLDOZED
output_directory: output
extracted_slice_parameters_directory: SliceParametersNZ/SliceParametersExtracted.txt
code: rt
extent_x: 251.20000000000002
extent_y: 250.4
extent_zmax: 36.0
extent_zmin: 0.0
sim_duration: 81.965
flo: 0.25
nx: 628
ny: 626
nz: 90
sufx: _rt01-h0.400

```

Hossack (Type 4 REL)
```
mag: 5.640075317543397
centroidDepth: 4.50333209967908
MODEL_LAT: -38.320909217531025
MODEL_LON: 176.20798721843408
MODEL_ROT: 1.0
hh: 0.4
min_vs: 0.5
model_version: '2.06'
topo_type: BULLDOZED
output_directory: output
extracted_slice_parameters_directory: SliceParametersNZ/SliceParametersExtracted.txt
code: rt
extent_x: 251.20000000000002
extent_y: 250.4
extent_zmax: 36.0
extent_zmin: 0.0
sim_duration: 81.96000000000001
flo: 0.25
nx: 628
ny: 626
nz: 90
sufx: _rt01-h0.400

```

Reply:
> Can we please see numeric comparisons (for VM dimensions and simulation time) between old and new for type 1,2,4?

See above

Reply:
@jasonmotha Removed hdepth, output_directory, extracted_slice_parameters_directory and code